samza-core/src/main/scala/org/apache/samza/container/SamzaContainer.scala

```diff
@@ -803,21 +806,40 @@ class SamzaContainer(
 
   def addShutdownHook {
     val runLoopThread = Thread.currentThread()
-    Runtime.getRuntime().addShutdownHook(new Thread() {
+    shutdownHookThread = new Thread("CONTAINER-SHUTDOWN-HOOK") {
       override def run() = {
         info("Shutting down, will wait up to %s ms" format shutdownMs)
         runLoop match {
           case runLoop: RunLoop => runLoop.shutdown
           case asyncRunLoop: AsyncRunLoop => asyncRunLoop.shutdown()
         }
-        runLoopThread.join(shutdownMs)
-        if (runLoopThread.isAlive) {
-          warn("Did not shut down within %s ms, exiting" format shutdownMs)
-        } else {
-          info("Shutdown complete")
+        try {
+          runLoopThread.join(shutdownMs)
+        } catch {
+          case e: Exception => // Ignore to avoid deadlock with uncaughtExceptioHandler. See SAMZA-1220
```

Comment:
IMHO it makes sense to catch, log and ignore all `Throwable` here.

- OOM errors happening in shutdown hook are usually in-actionable.
- We will bound the wait time by `shutdownMs` anyways.
- Other throwables happening in the `shutdownHook` will end up calling the `UncaughtExceptionHandler` resulting in a potential deadlock. Catching all `Throwable`s will prevent this.